File: app/src/main/java/com/gmail/ivan/synopsis/mvp/presenter/ThesisDetailsPresenter.java
package com.gmail.ivan.synopsis.mvp.presenter;

import android.os.AsyncTask;
import android.util.Log;

import com.gmail.ivan.synopsis.data.database.AppDataBase;
import com.gmail.ivan.synopsis.data.entity.Thesis;
import com.gmail.ivan.synopsis.mvp.contracts.ThesisDetailsContract;

import java.util.Objects;
import java.util.concurrent.ExecutionException;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public class ThesisDetailsPresenter
        extends BasePresenter<ThesisDetailsContract.View, ThesisDetailsContract.Router>
        implements ThesisDetailsContract.Presenter {

    private static final String TAG = ThesisDetailsPresenter.class.getSimpleName();

    @NonNull
    private final AppDataBase dataBase;

    @Nullable
    private Thesis thesis;

    public ThesisDetailsPresenter(@NonNull ThesisDetailsContract.Router router,
                                  @NonNull AppDataBase dataBase) {
        super(router);
        this.dataBase = dataBase;
    }

    @Override
    public void loadThesis(int thesisId) {
        try {
            thesis = new LoadThesisTask(dataBase).execute(thesisId)
                                                 .get();
            Objects.requireNonNull(getView())
                   .loadThesis(Objects.requireNonNull(thesis));
        } catch (ExecutionException | InterruptedException e) {
            Log.e(TAG, "loadThesis: ", e);
        }
    }

    @Override
    public void showEdit() {
        Objects.requireNonNull(getView())
               .showEditThesis();
    }

    @Override
    public void saveChanges() {
        new UpdateThesisTask(dataBase).execute(thesis);
        Objects.requireNonNull(getView())
               .showThesisDetails();
    }

    @Override
    public void showConfirmChanges() {
        getRouter().showConfirmChanges();
    }

    private static class UpdateThesisTask extends AsyncTask<Thesis, Void, Void> {

        @NonNull
        private final AppDataBase dataBase;

        UpdateThesisTask(@NonNull AppDataBase dataBase) {
            this.dataBase = dataBase;
        }

        @Override
        protected Void doInBackground(Thesis... theses) {
            dataBase.thesisRepository()
                    .updateThesis(theses[0]);
            return null;
        }
    }

    private static class LoadThesisTask extends AsyncTask<Integer, Void, Thesis> {

        @NonNull
        private final AppDataBase dataBase;

        LoadThesisTask(@NonNull AppDataBase dataBase) {
            this.dataBase = dataBase;
        }

        @Override
        protected Thesis doInBackground(Integer... integers) {
            return dataBase.thesisRepository()
                           .getThesis(integers[0]);
        }
    }
}

File: app/src/main/java/com/gmail/ivan/synopsis/mvp/presenter/ThesisDetailsPresenter.kt
package com.gmail.ivan.synopsis.mvp.presenter

import com.gmail.ivan.synopsis.data.entity.Thesis
import com.gmail.ivan.synopsis.data.repository.ThesisRepository
import com.gmail.ivan.synopsis.mvp.contracts.ThesisDetailsContract
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class ThesisDetailsPresenter(
    router: ThesisDetailsContract.Router,
    private val thesisRepository: ThesisRepository,
) : BasePresenter<ThesisDetailsContract.View, ThesisDetailsContract.Router>(router),
    ThesisDetailsContract.Presenter {

    private var thesis: Thesis? = null

    override fun loadThesis(thesisId: Int, isNewThesis: Boolean) {
        launch {
            thesis = withContext(Dispatchers.IO) {
                thesisRepository.getThesis(thesisId)
            }
            view?.loadThesis(thesis!!)
            if (!isNewThesis) {
                saveChanges()
            }
        }
    }

    override fun showEdit() {
        view?.showEditThesis()
    }

    override fun saveChanges() {
        launch {
            withContext(Dispatchers.IO) {
                thesisRepository.updateThesis(thesis!!)
            }
            view?.showThesisDetails()
        }
    }

    override fun showConfirmChanges() {
        router.showConfirmChanges()
    }

    companion object {

        private val TAG = ThesisDetailsPresenter::class.java.simpleName
    }
}